### PR TITLE
[Merged by Bors] - Simplify generics for the `SystemParamFunction` trait

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

The `SystemParamFunction` (and `ExclusiveSystemParamFunction`) trait is very cumbersome to use, due to it requiring four generic type parameters. These are currently all used as marker parameters to satisfy rust's trait coherence rules.

### Example (before)

```rust
pub fn pipe<AIn, Shared, BOut, A, AParam, AMarker, B, BParam, BMarker>(
    mut system_a: A,
    mut system_b: B,
) -> impl FnMut(In<AIn>, ParamSet<(AParam, BParam)>) -> BOut
where
    A: SystemParamFunction<AIn, Shared, AParam, AMarker>,
    B: SystemParamFunction<Shared, BOut, BParam, BMarker>,
    AParam: SystemParam,
    BParam: SystemParam,
```

## Solution

Turn the `In`, `Out`, and `Param` generics into associated types. Merge the marker types together to retain coherence.

### Example (after)

```rust
pub fn pipe<A, B, AMarker, BMarker>(
    mut system_a: A,
    mut system_b: B,
) -> impl FnMut(In<A::In>, ParamSet<(A::Param, B::Param)>) -> B::Out
where
    A: SystemParamFunction<AMarker>,
    B: SystemParamFunction<BMarker, In = A::Out>,
```

---

## Changelog

+ Simplified the `SystemParamFunction` and `ExclusiveSystemParamFunction` traits.

## Migration Guide

*The guide for #7023 has been merged into this one.*

For the `SystemParamFunction` trait, the type parameters `In`, `Out`, and `Param` have been turned into associated types.

```rust
// Before
fn my_generic_system<T, In, Out, Param, Marker>(system_function: T)
where
    T: SystemParamFunction<In, Out, Param, Marker>,
    T: Param: SystemParam,
{ ... }

// After
fn my_generic_system<T, Marker>(system_function: T)
where
    T: SystemParamFunction<Marker>,
{ ... }
```

For the `ExclusiveSystemParamFunction` trait, the type parameter `Param` has been turned into an associated type.
Also, `In` and `Out` associated types have been added, since exclusive systems now support system piping.

```rust
// Before
fn my_exclusive_system<T, Param, Marker>(system_function: T)
where
    T: ExclusiveSystemParamFunction<Param, Marker>,
    T: Param: ExclusiveSystemParam,
{ ... }

// After
fn my_exclusive_system<T, Marker>(system_function: T)
where
    T: ExclusiveSystemParamFunction<Marker>,
{ ... }
```

### CODE REVIEW

File: crates/bevy_ecs/src/system/exclusive_function_system.rs [hunk 1]
```diff
@@ -165,27 +158,31 @@ where
 ///
 /// This trait can be useful for making your own systems which accept other systems,
 /// sometimes called higher order systems.
-pub trait ExclusiveSystemParamFunction<In, Out, Param: ExclusiveSystemParam, Marker>:
-    Send + Sync + 'static
-{
+pub trait ExclusiveSystemParamFunction<Marker>: Send + Sync + 'static {
+    type In;
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/exclusive_function_system.rs:162]: Docs on these associated types would be very nice while we're here.

File: crates/bevy_ecs/src/system/function_system.rs [hunk 2]
```diff
@@ -606,19 +593,25 @@ where
 /// ```
 /// [`PipeSystem`]: crate::system::PipeSystem
 /// [`ParamSet`]: crate::system::ParamSet
-pub trait SystemParamFunction<In, Out, Param: SystemParam, Marker>: Send + Sync + 'static {
-    fn run(&mut self, input: In, param_value: SystemParamItem<Param>) -> Out;
+pub trait SystemParamFunction<Marker>: Send + Sync + 'static {
+    type In;
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/function_system.rs:597]: Docs here too.

File: crates/bevy_ecs/src/system/function_system.rs [hunk 3]
```diff
@@ -606,19 +593,25 @@ where
...
+    type Out;
+    type Param: SystemParam;
+    fn run(&mut self, input: Self::In, param_value: SystemParamItem<Self::Param>) -> Self::Out;
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/function_system.rs:600]: Also on run in both cases :)

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Much nicer. I think this is both clearer and less cumbersome.

More docs please though while we're here?

**Review by alice-i-cecile — APPROVED** [review]

**Review by james7132 — APPROVED** [review]

**Pull request closed without merging**